### ISSUE TITLE
[CA1873] False positives when log message wrapped in `IsEnabled()` check

### ISSUE THREAD

**Issue by martincostello** [dotnet/roslyn-analyzers#7690]
### Analyzer

**Diagnostic ID**: [CA1873](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/caA1873): `Evaluation of this argument may be expensive and unnecessary if logging is disabled`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 10 SDK](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: `SDK 10.0.100-preview.6.25311.101`

### Describe the bug

CA1873 fires even when an expensive logging expression is moved inside a block with a `if (logger.IsEnabled(LogLevel.))` condition.

For example:

```csharp
if (logger.IsEnabled(LogLevel.Information))
{
    Log.SomeInformationalLogUsingTheSourceGenerator(
        logger,
        [.. someDictionary.Select((p) => p.Key)]);
}
```

### Steps To Reproduce

1. Clone https://github.com/martincostello/costellobot/commit/0234077f6fe0bee92058404b4cf05d5173452a52
2. Run `build.ps1` in the root of the repository

### Expected behavior

The diagnostic is not raised as the operation is guarded by a check on the log level.

### Actual behavior

The diagnostic is raised.

## Additional context

Found through testing daily builds of .NET 10.

**Comment by mpidash** [comment]
Thanks for reporting this false positive and providing steps to reproduce. I was also able to do so locally with a unit test (using .NET 9).

Guarded log calls with compile-time source generated messages are detected, but **only** if used as an extension method. The other ways, as a static method (your repro) or as an instance method, are not currently covered right now, which was an oversight on my part when implementing the analyzer. Oops!

I will take a look at these additional cases this week.

**Comment by willdean** [comment]
In RC1 we're also seeing it report these for logger calls which have no expensive arguments at all - e.g. a string literal plus a string local as an argument.

**Comment by burnchar** [comment]
CA1873 also triggers even within an extension method in some cases, such as when passing the log level as a parameter to logger.Log:

<img width="1115" height="610" alt="Image" src="https://github.com/user-attachments/assets/c1c26837-20be-4813-907e-951e15a12cf2" />

```csharp
	public static void LogEnv(this ILogger logger, string? message = null, LogLevel level = LogLevel.Information)
	{
		if(logger.IsEnabled(LogLevel.Debug)) {
			logger.Log(
				level,
				"Machine:{MachineName}, OS:{OSVersion}, ProcessorCount:{ProcessorCount}, PID:{ProcessId}, User:{UserName}, WorkingDirectory:{CurrentDirectory}, CommandLine:{CommandLine}, ProcessPath:{ProcessPath}, Message:{Message}",
				Environment.MachineName, Environment.OSVersion.VersionString, Environment.ProcessorCount, Environment.ProcessId, Environment.UserName, Environment.CurrentDirectory,
				Environment.CommandLine, Environment.ProcessPath, message);
		}
	}
```

**Comment by willdean** [comment]
Still a problem in RTM:

```
  if (RunImmediately)
        {
            _logger.Log(LogLevel.Debug, "Task {Description} immediate run", _description);
            RunImmediately = false;
            return true;
        }
```

Creates a warning, even though _description is a private readonly string.

**Comment by PapaNappa** [comment]
[> In RC1 we're also seeing it report these for logger calls which have no expensive arguments at all - e.g. a string literal plus a string local as an argument.](https://github.com/dotnet/roslyn-analyzers/issues/7690#issuecomment-3321196229)

[> Creates a warning, even though _description is a private readonly string.](https://github.com/dotnet/roslyn-analyzers/issues/7690#issuecomment-3534015427)

[> I'm wondering, isn't this a false positive in general because a lot of non-MS loggers (like Serilog) do their checks internally, so there is no additional check required.](https://github.com/dotnet/roslyn-analyzers/issues/7690#issuecomment-3569731514)

Should we create a separate issue for these concerns? As this issue here originally is about not catching certain log guards.

Whereas a lot of comments are about the fact that the analysis might be a bit too sensitive in general.
